crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,425 @@
+///   there was a recent successful write, the next write will wait at least `COOLDOWN_INTERVAL`,
+///   and if there was a recent unsuccessful write, the next write will wait at least
+///   `RETRY_INTERVAL`. This is to prevent one committer from hogging the database.
+///
+/// - Number of pending rows. If this exceeds `H::BATCH_SIZE`, the committer will attempt to write
+///   out at most `H::CHUNK_SIZE` worth of rows to the DB.
+///
+/// If a write fails, the commiter will save the batch it meant to write and try to write them
+/// again at the next opportunity, potentially adding more rows to the batch if more have arrived
+/// in the interim.
+///
+/// This task will shutdown if canceled via the `cancel` token, or if the channel it receives data
+/// on has been closed by the handler for some reason.
+fn committer<H: Handler + 'static>(
+    db: Db,
+    mut rx: mpsc::Receiver<Indexed<H>>,
+    config: CommitterConfig,
+    metrics: Arc<IndexerMetrics>,
+    cancel: CancellationToken,
+) -> JoinHandle<()> {
+    spawn_monitored_task!(async move {
+        // The `poll` interval controls the maximum time to wait between commits, regardless of the
+        // amount of data available.
+        let mut poll = interval(config.commit_interval);
+        let mut cool = interval(COOLDOWN_INTERVAL);
+
+        // We don't care about keeping a regular cadence -- these intervals are used to guarantee
+        // things are spaced at out relative to each other.
+        poll.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        cool.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        // Buffer to gather the next batch to write. This may be non-empty at the top of a tick of
+        // the commiter's loop if the previous attempt at a write failed. Attempt is incremented
+        // every time a batch write fails, and is reset when it succeeds.
+        let mut attempt = 0;
+        let mut batch = vec![];
+
+        // Data for checkpoints that haven't been written yet. Note that `pending_rows` includes
+        // rows in `batch`.
+        let mut pending: BTreeMap<u64, Vec<H::Value>> = BTreeMap::new();
+        let mut pending_rows = 0;
+
+        // TODO: Track the watermark (keep track of out-of-order commits, and update the watermark
+        // table when we accumulate a run of contiguous commits starting from the current watermark).
+
+        loop {
+            tokio::select! {
+                // Break ties in favour of operations that reduce the size of the buffer.
+                biased;
+
+                _ = cancel.cancelled() => {
+                    info!(pipline = H::NAME, "Shutdown received, stopping committer");
+                    break;
+                }
+
+                // Time to write out another batch of rows, if there are any.
+                _ = poll.tick(), if pending_rows > 0 => {
+                    let Ok(mut conn) = db.connect().await else {
```

Comment:
What is the difference between connect() on Db and the get() on the connection pool?

Reply:
They are the same -- the `db: Db` is an abstraction over an async connection pool, and `db.connect()` is implemented in terms of `pool.get()`. I just called it `connect` because `db.get()` seemed confusing.